### DISCUSSION TITLE
The in-memory data format for duckDB

### DISCUSSION THREAD

**Discussion by deepuak** [duckdb/duckdb#7309] [General]
Hi,
Can someone please explain the in-memory format for duckDB in following scenarios?
1.) Read parquet files from folder and load to duckDB
2.) Create a duckDB database and load parquet files into this database

Also when loading parquet files or loading of duckDB database, will duckDB load entire data into memory or it pushes down predicates and only filter the required data from disk?

Regards,
DK.

**Comment by djouallah** [comment]
DuckDB will always try to pushes down predicates and only filter the required data from disk, either when querying directly parquet or the native storage format.

**Comment by deepuak** [comment]
Thanks. Could you also help me understand the following questions?

1. When data is stored in .duckdb format on disk, is it an in-memory format like arrow or on-disk format like parquet?
2. Any de-serialization happen when data is loaded from .duckdb to memory? 
3. Also when data is loaded from either .duckdb or parquet format from disk, is it converted to in-memory columnar format like arrow?

**Reply by djouallah** [reply]
Duckdb don't use arrow internally, that's my understanding anyway, you may find a better understanding here, I am just a user
https://www.youtube.com/watch?v=bZOvAKGkzpQ

**Comment by deepuak** [comment]
Thank you @djouallah that video was very useful.
Can duckDB manage with less memory? Say if my file size (paquet) is in TB can duckDB manage processing with couple GB of RAM?

**Reply by Alex-Monahan** [reply]
Yes, DuckDB is a streaming engine, meaning that it only pulls data into memory if it is needed for the current query, and it removes that data from memory when it is no longer needed. In most cases it can also page out to disk if intermediate results are too large for RAM (even if the initial data was small enough to fit). I've worked with a 450GB DuckDB on a machine with less RAM than that!

**Reply by deepuak** [reply]
Thanks you very much ! 
Does DuckDB memory map the file and work outside of JVM in the case of Java and can we use very less Heap Size in Java?
Also is there an ideal Heap memory size based on some formula?

**Comment by deepuak** [comment]
Also if someone can help understand how duckDB which is created in C++ is able to interact with say Java running in the same process?

**Reply by Alex-Monahan** [reply]
We have a JDBC connector and it does live in the same process where you are running Java.
